### PR TITLE
Iox #415 Add generic constructor to `cxx::variant`

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

* Add `variant(T&& arg) noexcept;`
* Rename/ rework some of the test cases

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes partly #415

### CODE REVIEW

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp [hunk 1]
```diff
@@ -128,26 +129,48 @@ bool variant_Test::DTorTest::dtorWasCalled = false;
...
+    iox::cxx::variant<ComplexClass, float> sut2{42.42F};
+
+    EXPECT_THAT(*sut2.get<float>(), Eq(42.42F));
+    EXPECT_THAT(sut2.index(), Eq(1U));
+}
+
+TEST_F(variant_Test, CreatingVariantFromLValueReturnsProvidedValueAndCorrectIndex)
+{
+    std::string string("Buhh");
+    iox::cxx::variant<std::string, float> sut2{string};
```
Comment (mossmaurice) [inline iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:154]: A `cxx::variant<std::string, float> sut2{"Buhh"}` is not possible because the type `char[]` is directly passed to `emplace<char[]>` which checks the validity of the type without considering implicit conversions. IMHO we don't need this, happy to add it if we find it useful.

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp [hunk 2]
```diff
@@ -128,26 +129,48 @@ bool variant_Test::DTorTest::dtorWasCalled = false;
...
+    sut.emplace<float>(1231.22F);
+    EXPECT_THAT(sut.index(), Eq(1U));
+}
+
+TEST_F(variant_Test, CreatingVariantFromPlainTypeReturnsProvidedValueAndCorrectIndex)
+{
+    iox::cxx::variant<ComplexClass, float> sut2{42.42F};
+
+    EXPECT_THAT(*sut2.get<float>(), Eq(42.42F));
+    EXPECT_THAT(sut2.index(), Eq(1U));
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:148]: Shouldn't this be an `ASSERT_THAT` and be done before the check for the value?

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:154]: I think for such cases the ctor could be called explicitly like `iox::cxx::variant<std::string, float> sut2{std::String{"Meh"}};` ... btw, would `iox::cxx::variant<std::string, float> sut2{{"Yay"}};`  work?

I guess this is not only related to strings but also to other types like having a `Foo(int, int)`?

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp [hunk 3]
```diff
@@ -128,26 +129,48 @@ bool variant_Test::DTorTest::dtorWasCalled = false;
...
+}
+
+TEST_F(variant_Test, CreatingVariantFromLValueReturnsProvidedValueAndCorrectIndex)
+{
+    std::string string("Buhh");
+    iox::cxx::variant<std::string, float> sut2{string};
+    EXPECT_THAT(sut2.get<std::string>()->c_str(), StrEq("Buhh"));
+    EXPECT_THAT(sut2.index(), Eq(0U));
+}
+
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:158]: Could you please add a test for `cxx::expected<float, float>`?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/variant.inl [hunk 4]
```diff
@@ -130,7 +140,7 @@ void variant<Types...>::call_element_destructor() noexcept
 template <typename... Types>
 template <typename T>
 inline typename std::enable_if<!std::is_same<T, variant<Types...>&>::value, variant<Types...>>::type&
-variant<Types...>::operator=(T&& f_rhs) noexcept
+variant<Types...>::operator=(T&& rhs) noexcept
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/variant.inl:143]: While I appreciate the removal of the `f` prefix, it adds quite some noise if it was used that massively like here. Please consider to do it in a separate PR it there are so many occurrences.

Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/variant.inl:143]: [Mimimimimi](https://youtu.be/_goMQolXcbs?t=19) :stuck_out_tongue: 

I'll consider it next time, but remember being a reviewer is not always a picnic :upside_down_face:

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:154]: Nope, `sut2{{"Buhh"}}` does not work. The template type deduction in C++14 is still limited: `couldn't deduce template parameter ‘T’`. `sut2{std::String{"Buhh"}}` works. It's like that for all types.

File: iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_cxx_variant.cpp:158]: I'm sure you meant `variant<float, float>` ;) Added.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp [hunk 5]
```diff
@@ -124,12 +147,23 @@ class variant
...
     /// @param[in] type type which should be created inside the variant
     /// @param[in] args variadic list of arguments which will be forwarded to the constructor to
     ///                 the type
     template <typename T, typename... CTorArguments>
     variant(const in_place_type<T>& type, CTorArguments&&... args) noexcept;
 
+    /// @brief creates a variant from a user supplied value
+    /// @tparam[in] T type of the value to be stored in the variant
+    /// @param[in] arg arg to be forwared to the c'tor of T
+    template <typename T,
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp:161]: Could you try to make this constructor constexpr.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp [hunk 6]
```diff
@@ -54,6 +55,26 @@ struct in_place_type
     using type = T;
 };
 
+template <typename N>
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp:58]: This is a construct which should be never used directly by the user. Could you therefore please move `is_in_place_index` and `is_in_place_type` to `variant_internal.hpp` into the `internal` namespace.

**Review by elfenpiff — COMMENTED** [review]
This is great - the variant is now a little bit cleaner and has another nice feature.

Thank you!

**Review by elBoberido — APPROVED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp (hunk 5 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/variant.hpp:141]: Good catch! Made all the c'tors and assignment ops `constexpr` like in the STL.

**Review by elfenpiff — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
